### PR TITLE
[feat]: [CDS-87082]: Add feature flag to support different infrastructure during environment propagation

### PR DESCRIPTION
**Pull request by lovish1999**
## Describe your changes
Add feature flag to support different infrastructure during environment propagation

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56878)

### CODE REVIEW

**Comment by lovish1999** [comment]
trigger commonchecks

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger ut0

**Comment by lovish1999** [comment]
trigger compile

**Comment by lovish1999** [comment]
trigger codebasehash

**Comment by lovish1999** [comment]
trigger pmd

**Comment by lovish1999** [comment]
trigger ut0

**Comment by lovish1999** [comment]
trigger ut0

**Review by yogesh-chauhan — APPROVED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/filters/DeploymentStageFilterJsonCreatorV2.java [hunk 1]
```diff
@@ -334,30 +334,34 @@ private void addFiltersFromEnvironment(FilterCreationContext filterCreationConte
...
+        } catch (Exception ex) {
+          log.warn(
+              format(
+                  "Exception occurred while saving filters for environment propagated from stage [%s] but having a different infrastructure selected.",
+                  env.getUseFromStage().getStage()),
+              ex);
+        }
       }
       return;
     }
```
Comment (tchaurasiya) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/creator/filters/DeploymentStageFilterJsonCreatorV2.java:367]: if this new FF is disabled, and environmentYaml contains useFromStage, Wont that throw exception as per below code

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/filters/DeploymentStageFilterJsonCreatorV2.java [hunk 2]
```diff
@@ -334,30 +334,34 @@ private void addFiltersFromEnvironment(FilterCreationContext filterCreationConte
...
-      } catch (Exception ex) {
-        log.warn(
-            format(
-                "Exception occurred while saving filters for environment propagated from stage [%s] but having a different infrastructure selected.",
-                env.getUseFromStage().getStage()),
-            ex);
+      if (ngFeatureFlagHelperService.isEnabled(filterCreationContext.getSetupMetadata().getAccountId(),
+              FeatureName.CDS_SUPPORT_DIFFERENT_INFRA_DURING_ENV_PROPAGATION)) {
+        try {
+          EnvironmentYamlV2 useFromStageEnvironmentYaml = getUseFromStageEnvironmentYaml(filterCreationContext, env);
```
Comment (tchaurasiya) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/creator/filters/DeploymentStageFilterJsonCreatorV2.java:341]: before calling this method shouldn't we check if env is using useFromStage?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/v1/DeploymentStagePlanCreator.java [hunk 3]
```diff
@@ -564,8 +564,8 @@ private String addServiceNode(PlanCreationContext ctx, YamlField specField,
     }
     String serviceNodeId = service.getUuid();
     ServiceDefinitionType serviceType = serviceEntityHelper.getServiceDefinitionTypeFromService(ctx, service);
-    planCreationResponseMap.putAll(ServiceAllInOnePlanCreatorUtils.addServiceNode(
-        specField, kryoSerializer, service, environment, serviceNodeId, nextNodeId, serviceType, envGroupRef, ctx));
+    planCreationResponseMap.putAll(ServiceAllInOnePlanCreatorUtils.addServiceNode(specField, kryoSerializer, service,
+        environment, serviceNodeId, nextNodeId, serviceType, envGroupRef, ctx, true));
```
Comment (tchaurasiya) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/stage/v1/DeploymentStagePlanCreator.java:568]: nit: Though this piece of code is not in use right now, But lets make last parameter false for now. we will onboard this feature to yaml v1 later

**Review by tchaurasiya — CHANGES_REQUESTED** [review]

**Review by tchaurasiya — APPROVED** [review]

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger ut5

**Comment by lovish1999** [comment]
trigger ut6

**Comment by lovish1999** [comment]
trigger ut7

**Comment by lovish1999** [comment]
trigger ut8

**Comment by lovish1999** [comment]
trigger ut9

**Comment by naidu-harness** [comment]
Review happened in other PR. Merging

**Pull request merged**